zerver/views/tusd.py

```diff
@@ -0,0 +1,65 @@
+import json
+import os
+from typing import Any, Dict
+from django.conf import settings
+from django.http import HttpRequest, HttpResponse
+from django.utils.translation import gettext as _
+from django.views.decorators.csrf import csrf_exempt
+from zerver.lib.exceptions import JsonableError
+from zerver.lib.response import json_success
+from zerver.lib.upload import check_upload_within_quota
+from zerver.lib.upload.base import create_attachment
+from shutil import move
+from zerver.models import UserProfile
+
+
+def handle_pre_create_hook(
+    request: HttpRequest, user_profile: UserProfile, data: Dict[str, Any]
+) -> HttpResponse:
+    file_size = data["Size"]
+    assert file_size != 0
+    if settings.MAX_FILE_UPLOAD_SIZE * 1024 * 1024 < file_size:
+        raise JsonableError(
+            _("Uploaded file is larger than the allowed limit of {} MiB").format(
+                settings.MAX_FILE_UPLOAD_SIZE,
+            )
+        )
+    check_upload_within_quota(user_profile.realm, file_size)
+    return json_success(request)
+
+
+def handle_pre_finish_hook(
+    request: HttpRequest, user_profile: UserProfile, data: Dict[str, Any]
+) -> HttpResponse:
+    file_id = data["ID"]
+    realm = user_profile.realm
+    file_name = data["MetaData"]["filename"]
+    file_size = data["Size"]
+    path_id = os.path.join(str(realm.id), file_id, file_name)
+    file_path = os.path.join(settings.LOCAL_FILES_DIR, path_id)
+    current_path = os.path.join(settings.LOCAL_UPLOADS_DIR, file_id)
+
+    os.makedirs(os.path.dirname(file_path), exist_ok=True)
+
+    move(current_path, file_path)
+
+    create_attachment(file_name, path_id, user_profile, realm, file_size)
+
+    return json_success(request)
+
+
+@csrf_exempt
+def handle_tusd_hook(
+    request: HttpRequest,
+    user_profile: UserProfile,
+) -> HttpResponse:
+    hook_name = request.headers.get("Hook-Name")
+
+    data = json.loads(request.body.decode("utf-8"))
+    data = data["Upload"]
+
+    if hook_name == "pre-create":
+        return handle_pre_create_hook(request, user_profile, data)
+    if hook_name == "pre-finish":
+        return handle_pre_finish_hook(request, user_profile, data)
+    return JsonableError("unexpected hook.")
```

Comment:
Can you write some tests for this code path? I think it'd be pretty helpful for verifying the logic/behavior. 

I guess I'm not entirely sure how to do that -- do we eventually need to change all the existing file upload tests to use this code path? For this stage of prototyping, probably just changing a basic test class in each one might be good enough.